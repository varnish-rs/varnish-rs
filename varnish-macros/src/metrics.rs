use std::collections::BTreeMap;
use std::fmt::Display;
use std::mem::size_of;
use std::str::FromStr;
use std::sync::atomic::AtomicU64;

use proc_macro2::{Ident, Literal, TokenStream};
use quote::quote;
use serde::Serialize;
use syn::meta::ParseNestedMeta;
use syn::punctuated::Punctuated;
use syn::spanned::Spanned;
use syn::token::Comma;
use syn::{Data, DeriveInput, Field, Fields, Type};

use crate::errors::Errors;
use crate::parser_utils::{find_attr, has_attr, parse_doc_str};
use crate::ProcResult;

type FieldList = Punctuated<Field, Comma>;

#[derive(Serialize, Clone)]
enum CType {
    #[serde(rename = "uint64_t")]
    Uint64,
}

impl CType {
    const fn size(&self) -> usize {
        match self {
            CType::Uint64 => size_of::<AtomicU64>(),
        }
    }
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "lowercase")]
enum MetricType {
    Counter,
    Gauge,
}

#[derive(Serialize, Clone, Default)]
#[serde(rename_all = "lowercase")]
enum Level {
    #[default]
    Info,
    Diag,
    Debug,
}

impl FromStr for Level {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "info" => Ok(Level::Info),
            "diag" => Ok(Level::Diag),
            "debug" => Ok(Level::Debug),
            _ => Err(format!(
                "Invalid level value '{s}'. Must be one of: info, diag, debug"
            )),
        }
    }
}

#[derive(Serialize, Clone, Default)]
#[serde(rename_all = "lowercase")]
enum Format {
    #[default]
    Integer,
    Bitmap,
    Duration,
    Bytes,
}

impl FromStr for Format {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "integer" => Ok(Format::Integer),
            "bitmap" => Ok(Format::Bitmap),
            "duration" => Ok(Format::Duration),
            "bytes" => Ok(Format::Bytes),
            _ => Err(format!(
                // TODO: can s come from an untrusted source? If so, we should not include it
                "Invalid format value '{s}'. Must be one of: integer, bitmap, duration, bytes"
            )),
        }
    }
}

#[derive(Serialize, Clone)]
struct VscMetricDef {
    pub name: String,
    #[serde(rename = "type")]
    pub metric_type: MetricType,
    pub ctype: CType,
    pub level: Level,
    pub oneliner: String, // "Counts the number of X", etc
    pub format: Format,
    pub docs: String,
    pub index: Option<usize>,
}

#[derive(Serialize)]
struct VscMetadata {
    version: &'static str,
    name: String,
    oneliner: String,
    order: u32,
    docs: String,
    elements: usize,
    // Using BTreeMap to ensure the order of elements
    elem: BTreeMap<String, VscMetricDef>,
}

pub fn derive_vsc_metric(input: &DeriveInput) -> ProcResult<TokenStream> {
    let name = &input.ident;

    if !has_repr_c(input) {
        return Err(syn::Error::new(
            name.span(),
            "VSC structs must be marked with #[repr(C)] for correct memory layout",
        )
        .into());
    }

    let fields = get_struct_fields(&input.data);
    validate_fields(fields)?;

    let metadata = generate_metadata_json(name, fields)?;
    let hashes = metadata
        .split('"')
        .skip(1) // Skip the first part which is before the first quote
        .map(|s| s.chars().take_while(|&c| c == '#').count() + 1)
        .max()
        .unwrap_or_default();
    let hashes = "#".repeat(hashes);
    let metadata = Literal::from_str(&format!("cr{hashes}\"{metadata}\"{hashes}")).unwrap();
    Ok(quote! {
        unsafe impl varnish::VscMetric for #name {
            fn get_metadata() -> &'static std::ffi::CStr {
                #metadata
            }
        }
    })
}

pub fn get_struct_fields(data: &Data) -> &FieldList {
    // FIXME: use errors collection
    match data {
        Data::Struct(data) => match &data.fields {
            Fields::Named(fields) => &fields.named,
            _ => panic!("Only named fields are supported"),
        },
        _ => panic!("Only structs are supported"),
    }
}

pub fn validate_fields(fields: &FieldList) -> ProcResult<()> {
    let mut errors = Errors::new();
    for field in fields {
        match &field.ty {
            Type::Path(path) => {
                let is_atomic_u64 = path
                    .path
                    .segments
                    .last()
                    .is_some_and(|seg| seg.ident == "AtomicU64");

                if !is_atomic_u64 {
                    let field_name = field.ident.as_ref().map(ToString::to_string).unwrap();
                    errors.push(syn::Error::new(
                        Spanned::span(field),
                        format!("Field {field_name} must be of type AtomicU64"),
                    ));
                }
            }
            _ => {
                errors.push(syn::Error::new(
                    Spanned::span(field),
                    "Field types must be of type AtomicU64",
                ));
            }
        }
    }
    Ok(errors.into_result()?)
}

fn generate_metrics(fields: &FieldList) -> ProcResult<BTreeMap<String, VscMetricDef>> {
    let mut offset = 0;
    let mut result = BTreeMap::new();

    for field in fields {
        let name = field.ident.as_ref().unwrap().to_string();

        let has_counter = has_attr(&field.attrs, "counter");
        let has_gauge = has_attr(&field.attrs, "gauge");
        let metric_type = if has_counter && has_gauge {
            // FIXME: use errors collection
            panic!("Field {name} cannot have both #[counter] and #[gauge] attributes");
        } else if has_counter {
            MetricType::Counter
        } else if has_gauge {
            MetricType::Gauge
        } else {
            // FIXME: use errors collection
            panic!("Field {name} must have either #[counter] or #[gauge] attribute")
        };

        let doc_str = parse_doc_str(&field.attrs);
        let mut doc_lines = doc_str.split('\n').filter(|s| !s.is_empty());
        let oneliner = doc_lines.next().unwrap_or_default().to_string();
        let docs = doc_lines.next().unwrap_or_default().to_string();

        let (level, format) = parse_metric_attributes(
            field,
            match metric_type {
                MetricType::Counter => "counter",
                MetricType::Gauge => "gauge",
            },
        )?;

        let ctype = CType::Uint64;
        let index = Some(offset);
        offset += ctype.size();

        result.insert(
            name.clone(),
            VscMetricDef {
                name,
                metric_type,
                ctype,
                level,
                oneliner,
                format,
                docs,
                index,
            },
        );
    }

    Ok(result)
}

pub fn generate_metadata_json(name: &Ident, fields: &FieldList) -> ProcResult<String> {
    let metrics = generate_metrics(fields)?;

    let metadata = VscMetadata {
        version: "1",
        name: name.to_string(),
        oneliner: format!("{name} statistics"),
        order: 100,
        docs: String::new(),
        elements: metrics.len(),
        elem: metrics,
    };

    Ok(
        serde_json::to_string(&metadata)
            .map_err(|e| syn::Error::new(name.span(), e.to_string()))?,
    )
}

#[expect(clippy::unnecessary_wraps)]
fn parse_metric_attributes(field: &Field, metric_type: &str) -> ProcResult<(Level, Format)> {
    /// Helper function to parse the attribute value
    fn parse<T: FromStr>(id: &Ident, meta: &ParseNestedMeta, field: &Field) -> Result<T, syn::Error>
    where
        <T as FromStr>::Err: Display,
    {
        let value = meta.value()?.parse::<syn::LitStr>()?.value();
        T::from_str(&value).map_err(|e| {
            meta.error(format!(
                "Invalid {id} value for field {}: {e}",
                field.ident.as_ref().unwrap()
            ))
        })
    }

    let mut level = Level::default();
    let mut format = Format::default();
    if let Some(attrs) = find_attr(&field.attrs, metric_type) {
        // FIXME: parse_nested_meta() returns an error:
        //     "expected attribute arguments in parentheses ..."
        //   ignoring seems to work, but that's a strange approach and needs fixing
        let _ = attrs.parse_nested_meta(|meta| {
            if let Some(ident) = meta.path.get_ident() {
                if ident == "level" {
                    level = parse(ident, &meta, field)?;
                } else if ident == "format" {
                    format = parse(ident, &meta, field)?;
                }
            }
            Ok(())
        });
    }
    Ok((level, format))
}

pub fn has_repr_c(input: &DeriveInput) -> bool {
    input.attrs.iter().any(|attr| {
        if !attr.path().is_ident("repr") {
            return false;
        }

        let Ok(meta) = attr.parse_args::<syn::Meta>() else {
            return false;
        };

        matches!(meta, syn::Meta::Path(path) if path.is_ident("C"))
    })
}
