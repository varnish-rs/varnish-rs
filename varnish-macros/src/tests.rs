use std::fs::read_to_string;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;

use insta::{assert_snapshot, with_settings};
use proc_macro2::TokenStream;
use quote::quote;
use regex::Regex;
use syn::{Data, DeriveInput, ItemMod, ItemStruct};

use crate::gen_docs::gen_doc_content;
use crate::generator::render_model;
use crate::metrics::derive_vsc_metric;
use crate::parser::tokens_to_model;
use crate::parser_utils::remove_attr;

static RE_VARNISH_VER: LazyLock<Regex> = LazyLock::new(|| {
    // Use regex to remove "Varnish 7.5.0 eef25264e5ca5f96a77129308edb83ccf84cb1b1" and similar.
    // Also removes any pre-builds and other versions because we assume a double-quote at the end.
    Regex::new(r"Varnish \d+\.[-+. 0-9a-z]+").unwrap()
});

static RE_JSON_BLOB: LazyLock<Regex> = LazyLock::new(|| {
    // Use regex to remove "const JSON: &CStr = c\"...\";"
    Regex::new(r#"const JSON: &CStr = c"([^\n]+)";\n"#).unwrap()
});

/// Read content of the ../../varnish/tests/pass directory that should also pass full compilation tests,
/// parse them and create snapshots of the model and the generated code.
#[test]
fn parse_pass_tests() {
    run_parse_tests("../varnish/tests/pass/*.rs");
}

#[test]
fn parse_pass_ffi_tests() {
    run_parse_tests("../varnish/tests/pass_ffi/*.rs");
}

fn run_parse_tests(path: &str) {
    let version = if cfg!(varnishsys_6) {
        "_v6"
    } else if cfg!(varnishsys_77_vmod_data) {
        ""
    } else {
        "_v76"
    };
    let snapshot_path = format!("../../varnish/snapshots{version}");
    with_settings!({ snapshot_path => snapshot_path, omit_expression => true, prepend_module_to_snapshot => false }, {
        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(path);
        for file in glob::glob(path.to_str().unwrap()).unwrap() {
            test_file(&file.unwrap());
        }
    });
}

fn test_file(filepath: &Path) {
    let file = filepath.to_str().unwrap();
    eprintln!("Processing file: {file}");
    let src = read_to_string(filepath).expect("unable to read file");
    let syntax = syn::parse_file(&src).expect("unable to parse file");
    let mut has_vmod = false;
    let mut has_metrics = false;
    for item in syntax.items {
        if let syn::Item::Mod(mut item) = item {
            assert!(!has_vmod, "Multiple vmod modules found in {file}");
            has_vmod = true;
            // FIXME: use this attribute as an arg for the test
            let _arg = remove_attr(&mut item.attrs, "vmod").unwrap();
            let name = format!(
                "{}_{}",
                filepath.file_stem().unwrap().to_string_lossy(),
                item.ident
            );
            test(&name, quote! {}, item);
            // FIXME: pass proper attribute info
            // test(&name, quote! { #arg }, item);
        } else if let syn::Item::Struct(item) = item {
            // Check if this has #[derive(VscMetric)]
            for attr in &item.attrs {
                // Check for derive attribute
                if let syn::Meta::List(meta_list) = &attr.meta {
                    if meta_list.path.is_ident("derive") {
                        let tokens = meta_list.tokens.to_string();
                        if tokens.contains("VscMetric") {
                            assert!(!has_metrics, "Multiple VscMetric structs found in {file}");
                            has_metrics = true;
                            let name = format!(
                                "{}_{}",
                                filepath.file_stem().unwrap().to_string_lossy(),
                                item.ident
                            );
                            test_metric(&name, item);
                            break;
                        }
                    }
                }
            }
        }
    }
    assert!(
        has_vmod || has_metrics,
        "No vmod modules or derive(VscMetric) found in file {file}"
    );
}

fn test_metric(name: &str, item: ItemStruct) {
    let ItemStruct {
        attrs,
        vis,
        struct_token,
        ident,
        generics,
        fields,
        semi_token,
    } = item;

    let input = DeriveInput {
        attrs,
        vis,
        ident,
        generics,
        data: Data::Struct(syn::DataStruct {
            struct_token,
            fields,
            semi_token,
        }),
    };

    let Ok(code) = derive_vsc_metric(&input).map_err(|err| {
        // On error, save the error output as a snapshot and return early.
        let err = err.into_compile_error();
        with_settings!({ snapshot_suffix => "error" }, { assert_snapshot!(name, err) });
    }) else {
        return;
    };
    let code = parse_gen_code(name, &code.to_string());
    with_settings!({ snapshot_suffix => "code" }, { assert_snapshot!(name, code) });
}

fn test(name: &str, args: TokenStream, mut item_mod: ItemMod) {
    let Ok(info) = tokens_to_model(args, &mut item_mod).map_err(|err| {
        // On error, save the error output as a snapshot and return early.
        let err = err.into_compile_error();
        with_settings!({ snapshot_suffix => "error" }, { assert_snapshot!(name, err) });
    }) else {
        return;
    };

    with_settings!({ snapshot_suffix => "model" }, { assert_snapshot!(name, format!("{info:#?}")) });
    with_settings!({ snapshot_suffix => "docs" }, { assert_snapshot!(name, gen_doc_content(&info)) });

    let file = render_model(item_mod, &info).to_string();
    let generated = parse_gen_code(name, &file);

    let mut code = RE_VARNISH_VER
        .replace_all(&generated, "Varnish (version) (hash)")
        .to_string();

    // Extract a JSON string and remove it from the original to avoid recording it twice
    let json = if let Some(v) = RE_JSON_BLOB.captures(code.as_ref()) {
        let res = v.get(1).unwrap().as_str().to_string();
        code = RE_JSON_BLOB
            .replace(
                code.as_ref(),
                "const JSON: &CStr = c\"(moved to @json.snap files)\";\n",
            )
            .to_string();
        res
    } else {
        String::new()
    };

    with_settings!({ snapshot_suffix => "code" }, { assert_snapshot!(name, code) });

    let json = &json
        .replace("\\\"", "\"")
        .replace("\\u{2}", "\u{2}")
        .replace("\\u{3}", "\u{3}")
        .replace("\\\\", "\\")
        // this is a bit of a hack because the double-escaping gets somewhat incorrectly parsed
        .replace("\\n", "\n");

    with_settings!({ snapshot_suffix => "json" }, { assert_snapshot!(name, json) });
}

fn parse_gen_code(name: &str, file: &str) -> String {
    let parsed = match syn::parse_file(file) {
        Ok(v) => v,
        Err(e) => {
            // We still save the generated code, but fail the test. Allows easier error debugging.
            with_settings!({ snapshot_suffix => "code" }, { assert_snapshot!(name, file) });
            panic!("Failed to parse generated code in test {name}: {e}");
        }
    };

    prettyplease::unparse(&parsed)
}
