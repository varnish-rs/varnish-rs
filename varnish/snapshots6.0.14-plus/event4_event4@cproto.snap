---
source: varnish-macros/src/tests.rs
---
struct Vmod_event4_Func {
  vmod_event_f *f_on_event;
};

static struct Vmod_event4_Func Vmod_event4_Func;
